pkg/intermediate/aggregate.go

```diff
@@ -148,31 +171,93 @@ func (a *AggregationProcess) ForAllRecordsDo(callback FlowKeyRecordMapCallBack)
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	currTime := time.Now()
+	if a.expirePriorityQueue.Len() > 0 {
+		topItem := a.expirePriorityQueue.Peek()
+		if topItem.activeExpireTime.Before(topItem.inactiveExpireTime) {
+			return topItem.activeExpireTime.Sub(currTime)
+		} else {
+			return topItem.inactiveExpireTime.Sub(currTime)
+		}
+	}
+	if a.activeExpiryTimeout < a.inactiveExpiryTimeout {
+		return a.activeExpiryTimeout
 	}
+	return a.inactiveExpiryTimeout
 }
 
-func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) error {
+func (a *AggregationProcess) ForAllExpiredFlowRecordsDo(callback FlowKeyRecordMapCallBack) error {
 	a.mutex.Lock()
 	defer a.mutex.Unlock()
-	_, exists := a.flowKeyRecordMap[flowKey]
-	if !exists {
-		return fmt.Errorf("flow key is not present in the map")
+
+	if a.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for a.expirePriorityQueue.Len() > 0 {
+		topItem := a.expirePriorityQueue.Peek()
+		if topItem.activeExpireTime.After(currTime) && topItem.inactiveExpireTime.After(currTime) {
+			// We do not have to check other items anymore.
+			break
+		}
+		// Pop the record item from the priority queue
+		pqItem := heap.Pop(&a.expirePriorityQueue).(*ItemToExpire)
+		if !pqItem.flowRecord.ReadyToSend {
+			// Reset the timeouts and add the record to priority queue.
+			// Delete the record after max retries.
+			pqItem.flowRecord.waitForReadyToSendRetries = pqItem.flowRecord.waitForReadyToSendRetries + 1
+			if pqItem.flowRecord.waitForReadyToSendRetries > MaxRetries {
+				klog.V(2).Infof("Deleting the record after waiting for ready to send with key: %v record: %v", pqItem.flowKey, pqItem.flowRecord)
+				if err := a.deleteFlowKeyFromMapWithoutLock(*pqItem.flowKey); err != nil {
+					return fmt.Errorf("error while deleting flow record after max retries: %v", err)
+				}
+			} else {
+				pqItem.activeExpireTime = currTime.Add(a.activeExpiryTimeout)
+				pqItem.inactiveExpireTime = currTime.Add(a.inactiveExpiryTimeout)
+				heap.Push(&a.expirePriorityQueue, pqItem)
+			}
+			continue
+		}
+		err := callback(*pqItem.flowKey, *pqItem.flowRecord)
+		if err != nil {
+			return fmt.Errorf("callback execution failed for popped flow record with key: %v, record: %v, error: %v", pqItem.flowKey, pqItem.flowRecord, err)
+		}
+		// Delete the flow record if it is expired because of inactive expiry timeout.
+		if pqItem.inactiveExpireTime.Before(currTime) {
+			if err = a.deleteFlowKeyFromMapWithoutLock(*pqItem.flowKey); err != nil {
+				return fmt.Errorf("error while deleting flow record after inactive expiry: %v", err)
+			}
+			continue
+		}
+		// Reset the expireTime for the popped item and push it to the priority queue.
+		if pqItem.activeExpireTime.Before(currTime) {
+			// Reset the active expire timeout and push the record into priority
+			// queue.
+			pqItem.activeExpireTime = currTime.Add(a.activeExpiryTimeout)
```

Comment:
@shihhaoli In that case, we could also have gone for a lower inactive timeout than the active timeout, where we delete the flow record as soon as the inactive timeout is elapsed A new record can be added when a new flow record comes in. 

We went this route to reduce memory allocations and cleanups for the case of long-lived flows. In addition, as Yiou mentioned this also benefits to have a continuous time-series when visualizing flow metrics.

Reply:
Thanks guys. Sorry a couple more questions.

So in the current logic, we could still export the record when inactive timeout expires because the code only checks if the popped record is ready to send without checking it is due to which timeout. Is that right?

Do we also have an attribute in the record storing the last updated time?
If so, maybe we can only use active timeout as the periodical interval for the timeout queue (i.e. priority queue). Whenever it expires, then check if the time between the last updated time and current timeout exceeds the inactive timeout value. If yes, we can stop there (i.e. no longer adding the record back to the queue). In this case, the data point will be updated regularly based on active timeout interval until inactive timeout expires.

Anyhow, this is just my thought based on my current understanding. It may not reflect what you intend to do. Thanks.

Reply:
>So in the current logic, we could still export the record when inactive timeout expires because the code only checks if the popped record is ready to send without checking it is due to which timeout. Is that right?

Yes, that is correct. This ReadyToSend is to wait for flow records to be completely correlated as per our expectations. We do not send any flow records in this scenario.

>Do we also have an attribute in the record storing the last updated time?

Yes, we do have that, but this is taken from the `flowEndSeconds` element in the flow record. This is added by Flow Exporter at Antrea Agent and not Flow Aggregator. With the logic you suggested, we cannot expire the flow record at a constant interval for all inactive flow records. It depends on the time instance when the last flow record has been received.

Thanks for the comments, Shih-hao.

Reply:
Got it. Thanks.
I was trying to see if we can avoid updating the queue due to change on inactive timeout because moving objects in the queue could be more expensive than just checking certain condition.